commons-test/src/main/java/org/infinispan/commons/test/RunningTestsRegistry.java

```diff
@@ -0,0 +1,96 @@
+package org.infinispan.commons.test;
+
+import static java.util.Collections.singleton;
+import static java.util.concurrent.TimeUnit.MINUTES;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.InputStreamReader;
+import java.lang.management.ManagementFactory;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+
+/**
+ * Keep track of running tests and interrupt them if they take more than {@link #MAX_TEST_SECONDS} seconds.
+ *
+ * @author Dan Berindei
+ * @since 9.2
+ */
+class RunningTestsRegistry {
+   private static final long MAX_TEST_SECONDS = MINUTES.toSeconds(5);
+
+   private static final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor(
+         r -> new Thread(r, "LongTestsWatcher"));
+   private static final Map<Thread, ScheduledFuture<?>> scheduledTasks = new ConcurrentHashMap<>();
+
+   static void unregisterThreadWithTest() {
+      ScheduledFuture<?> killTask = scheduledTasks.remove(Thread.currentThread());
+      killTask.cancel(true);
+   }
+
+   static void registerThreadWithTest(String testName, String simpleName) {
+      Thread testThread = Thread.currentThread();
+      ScheduledFuture<?> future = executor.schedule(
+         () -> killLongTest(testThread, testName, simpleName), MAX_TEST_SECONDS, SECONDS);
+      scheduledTasks.put(testThread, future);
+   }
+
+   @SuppressWarnings("deprecation")
+   private static void killLongTest(Thread testThread, String testName, String simpleName) {
+      try {
+         String safeTestName = testName.replaceAll("[^a-zA-Z0-9=]", "_");
+         System.err.printf(
+            "Test %s has been running for more than %d seconds. Interrupting the test thread and dumping " +
+               "thread stacks of the test suite process and its children.\n",
+            testName, MAX_TEST_SECONDS);
+
+         String jvmName = ManagementFactory.getRuntimeMXBean().getName();
```

Comment:
Wow, nice trick! Is it going to work if you disable JMX in Infinispan configuration?

Reply:
Why wouldn't it work? JMX here is just simply used to get a thread dump. There's nothing specific to Infinispan MBeans which are registered regardless of whether stats are enabled or not.

Reply:
Indeed, this uses the [platform MXBeans](https://docs.oracle.com/javase/7/docs/technotes/guides/management/mxbeans.html).